c/tskit/trees.c

```diff
@@ -4551,47 +4551,95 @@ tsk_tree_position_in_interval(const tsk_tree_t *self, double x)
     return self->interval.left <= x && x < self->interval.right;
 }
 
+static int
+tsk_tree_seek_from_null(tsk_tree_t *self, double x, tsk_flags_t TSK_UNUSED(options))
+{
+    int ret = 0;
+    tsk_size_t edge, num_edges = self->tree_sequence->tables->edges.num_rows,
+                     num_trees = self->tree_sequence->num_trees;
+    tsk_id_t p, c, e, tree_index;
+    const double L = tsk_treeseq_get_sequence_length(self->tree_sequence);
+    const tsk_id_t *restrict edge_parent = self->tree_sequence->tables->edges.parent;
+    const tsk_id_t *restrict edge_child = self->tree_sequence->tables->edges.child;
+    const double *restrict edge_left = self->tree_sequence->tables->edges.left;
+    const double *restrict edge_right = self->tree_sequence->tables->edges.right;
+    const double *restrict breakpoints = self->tree_sequence->breakpoints;
+    const tsk_id_t *restrict insertion
+        = self->tree_sequence->tables->indexes.edge_insertion_order;
+    const tsk_id_t *restrict removal
+        = self->tree_sequence->tables->indexes.edge_removal_order;
+
+    /* NOTE: this differ's from JK's proposal by not using
+     * edge diffs explicitly. However:
+     * 1. It benchmarked faster (in rust).
+     * 2. It is a lot less code and doesn't have
+     * to do "edge diffs, but backwards".
+     */
+    self->num_edges = 0;
+
+    if (x <= L / 2.0) {
+        for (edge = 0; edge < num_edges; ++edge) {
+            e = insertion[edge];
+            if (edge_left[e] > x) {
+                break;
+            }
+            if (x >= edge_left[e] && x < edge_right[e]) {
+                p = edge_parent[e];
+                c = edge_child[e];
+                tsk_tree_insert_edge(self, p, c, e);
+            }
+        }
+    } else {
+        for (edge = 0; edge < num_edges; ++edge) {
+            e = removal[num_edges - edge - 1];
+            if (edge_right[e] < x) {
+                break;
+            }
+            if (x >= edge_left[e] && x < edge_right[e]) {
+                p = edge_parent[e];
+                c = edge_child[e];
+                tsk_tree_insert_edge(self, p, c, e);
+            }
+        }
+    }
+
+    /* NOTE: this is where I'm a bit nervous */
+
+    /* We need to get the index of the tree in order to update other
+     * data fields. */
+    for (tree_index = 0; tree_index < (tsk_id_t) num_trees - 1; ++tree_index) {
+        if (breakpoints[tree_index + 1] > x) {
+            break;
+        }
+    }
+    self->index = tree_index;
+    self->interval.left = breakpoints[tree_index];
+    self->interval.right = breakpoints[tree_index + 1];
+    self->left_index = insertion[tree_index];
```

Comment:
This isn't right - you're still thinking of the per-tree arrays that you're storing. You need to store `e` here probably.

You need to keep track of the "removal" value too.

Reply:
Yes, this is where things are going wrong.  I am unlikely to get back to this until next week.

Reply:
Fixed.